app/hooks/use-countdown-timer.ts

```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState } from "react"
+
+type useCountdownTimerReturnValue = {
+  timeLeft: number
+  startCountdownTimer: (seconds: number, callback?: () => void) => void
+  stopCountdownTimer: () => void
+  resetCountdownTimer: (seconds: number, newCallback?: () => void) => void
+  resetCountdownTimerAndExecuteCallback: (
+    seconds: number,
+    newCallback?: () => void,
+  ) => void
+  stopCountdownTimerAndExecuteCallback: (newCallback?: () => void) => void
+}
+
+export const useCountdownTimer = (): useCountdownTimerReturnValue => {
+  const [timeLeft, setTimeLeft] = useState<number>()
+  const [completedCallback, setCompletedCallback] = useState<() => void>()
+
+  useEffect(() => {
+    if (timeLeft !== undefined) {
+      const interval = setInterval(() => {
+        setTimeLeft(timeLeft - 1)
+      }, 1000)
+
+      if (timeLeft === 0) {
+        clearInterval(interval)
+        if (completedCallback) {
+          completedCallback()
+        }
+      }
+
+      return () => {
+        clearInterval(interval)
+      }
+    }
+    // We need to return from all code paths but we only need a callback if an interval is set so returning undefined is fine
+    return undefined
+  }, [timeLeft, completedCallback])
```

Comment:
Should ```setTimeLeft``` be a part of the dependency list?

Reply:
Yeah agreed - have added it to the dependency list